install_kustomize: support linux/aarch64, with fallback to old behavior

cfstras opened the pull request:
In docker on an M1 Mac -- and probably also on any ARM linux machine, the value for `uname -m` is `aarch64`.
The install script only detects `arm64` (which is the ARM uname value for native darwin).

Since the default is amd64, this leads to the script **installing the wrong (amd64) version on ARM** docker, instead of the arm64 version.

---

Caveat:

I appreciate that some people might be using this script to install older versions (which might not have an arm64 build yet), and have qemu-binfmt-misc installed, so this approach worked for them (but sometimes crashes, at least on M1 machines, a different issue). That's why I added a fallback that goes back to the old behavior if the arm64 binary does not exist.

Also fixed some shellcheck issues.


Any comments and feedback appreciated :)

cfstras commented:
/label tide/merge-method-squash

Currently waiting from my org for CLA approval.

cfstras commented:
/assign knverey yuwenma

KnVerey commented on hack/install_kustomize.sh:
```diff
@@ -73,7 +73,17 @@ function readlink_f {
+  local opsys=$1
+  local arch=$2
+
+  echo "${releases}" |\
```
this is relying on `releases` being defined globally. Let's pass it in like you're doing for opsys and arch.

KnVerey commented on hack/install_kustomize.sh:
```diff
@@ -123,31 +133,38 @@ s390x)
+    # fallback to the old behavior of downloading amd64 binaries on aarch64 systems.
+    # People might have qemu-binfmt-misc installed, so it worked for them previously.
+    echo "Version $version does not exist for ${opsys}_arm64, trying ${opsys}_amd64 instead."
+    RELEASE_URL="$(find_release_url "$opsys" "amd64")"
```
This is currently always replacing the arm64 version. I think you wanted something like:

```bash
if [[ "$(uname -m)" == "aarch64" ]] && [[ -z "$RELEASE_URL" ]]; then
    # fallback to the old behavior of downloading amd64 binaries on aarch64 systems.
    # People might have qemu-binfmt-misc installed, so it worked for them previously.
    echo "Version $version does not exist for ${opsys}_arm64, trying ${opsys}_amd64 instead."
    arch=amd64
    RELEASE_URL="$(find_release_url "$opsys" "$arch")"
fi
```

cfstras replied:
Yes, indeed. Updated.

burmanm commented:
@cfstras Any chance you could update your script based on the comments to get this merged? Although I'd be happy with just:

```patch
112c112
< arm64)
---
> arm64 | aarch64)
```

Since it works just fine in Ubuntu 22.04.01. I'm not sure how important those old "qemu-hacks" are to support in the installation scripts, seems to add a lot of extra stuff just to maintain them.

burmanm commented:
@cfstras But there's still the "if [[ "$(uname -m)" == "aarch64" ]]; then" unfixed as mentioned in the comment. So nothing uses aarch64. Is this PR missing a commit?

cfstras commented:
@burmanm apologies, I didn't get to this until now. My last fixes seem not to have made the push, so I addressed all comments and did a round of testing now.
I'm using `$arch` directly now, which makes it a bit easier.

CC @KnVerey if you could take another look please? :)

KnVerey commented:
/cc @koba1t

koba1t commented:
I think this script is working well.
Thanks for your contribution!

/lgtm

KnVerey commented on hack/install_kustomize.sh:
```diff
@@ -123,31 +134,38 @@ s390x)
+    echo "Version $version does not exist for ${opsys}_arm64, trying ${opsys}_amd64 instead."
+    RELEASE_URL="$(find_release_url "$releases" "$opsys" "amd64")"
+    if [[ -n "$RELEASE_URL" ]]; then
+      arch=amd64
```
Since we've informed the user that we're about to try amd64, I think we should change this unconditionally before L150 and then not need to hard code it there. Otherwise the messaging is a little confusing:

```shell
# attempting to install v3.2.0 on an m1 mac
Version v3.2.0 does not exist for darwin_arm64, trying darwin_amd64 instead.
Version v3.2.0 does not exist or is not available for darwin/arm64. # so, did we actually try amd64 or not?
```

cfstras replied:
done :)

KnVerey commented on hack/install_kustomize.sh:
```diff
@@ -123,31 +134,38 @@ s390x)
+if [[ "$arch" == "arm64" ]] && [[ -z "$RELEASE_URL" ]] ; then
+    # fallback to the old behavior of downloading amd64 binaries on aarch64 systems.
+    # People might have qemu-binfmt-misc installed, so it worked for them previously.
+    echo "Version $version does not exist for ${opsys}_arm64, trying ${opsys}_amd64 instead."
```
nit: to match the formatting used at L157:

```suggestion
    echo "Version $version does not exist for ${opsys}/arm64, trying ${opsys}/amd64 instead."
```

cfstras replied:
done, and rebased onto latest master :) LMK if that fixes the CI.

KnVerey reviewed (commented):
Just two very minor requests, plus I think this needs a rebase to get CI to run (the approval button I need to click isn't even showing up). 🙏

KnVerey commented:
/ok-to-test

The Github Actions button also showed up after the rebase, and they're running now too!

KnVerey commented:
/lgtm
/approve
